### PR TITLE
Update project

### PR DESCRIPTION
**Pull request by FreeApophis**
Updated to the newest coding standards
* C#9 and .NET5
* Funcky 2.3 and other NuGets updated
* ReSharped

### CODE REVIEW

File: .github/workflows/dotnetcore.yml [hunk 1]
```diff
@@ -17,7 +17,7 @@ jobs:
     - name: Setup .NET Core
       uses: actions/setup-dotnet@v1
       with:
```
Comment (tautropfli) [inline .github/workflows/dotnetcore.yml]: You can remove `dotnet-version` (and the entire `with` section). `actions/setup-dotnet@v1` reads the version from `global.json`

File: FileManipulator.Test/GlobalJsonManipulatorTest.cs [hunk 2]
```diff
@@ -128,14 +128,14 @@ public async Task ThrowsWhenMsbuildSdksKeyHasIncorrectType()
...
         public static TheoryData<string, GlobalJsonModification> GetModificationsToNewFile()
-            => new TheoryData<string, GlobalJsonModification>
+            => new ()
             {
                 {
                     $"{{{NewLine}" +
                     $"    \"msbuild-sdks\": {{{NewLine}" +
                     $"        \"{MsBuildSdkNugetPackageId}\": \"{MsBuildSdkVersion}\"{NewLine}" +
                     $"    }}{NewLine}" +
-                    $"}}",
```
Comment (tautropfli) [inline FileManipulator.Test/GlobalJsonManipulatorTest.cs]: I think these `$`s are here for alignment purposes with the previous lines

File: FileManipulator.Test/NugetPackageSourcesTest.cs [hunk 3]
```diff
@@ -178,8 +179,8 @@ public void DisablePackageSource(string configFileContent, string expectedOutput
             Assert.Equal(expectedFileContent, ReadNugetTestFile(testEnvironment));
         }
 
-        private static INugetPackageSourceManipulator CreateNugetPackageSources(ILogger? logger = null)
-            => new NugetPackageSourceManipulator(() => logger ?? new Mock<ILogger>().Object);
+        private static INugetPackageSourceManipulator CreateNugetPackageSources(Option<ILogger> logger = default)
+            => new NugetPackageSourceManipulator(() => logger.GetOrElse(() => new Mock<ILogger>().Object));
```
Comment (tautropfli) [inline FileManipulator.Test/NugetPackageSourcesTest.cs]:
```suggestion
            => new NugetPackageSourceManipulator(() => logger.GetOrElse(() => Mock.Of<ILogger>()));
```

File: FileManipulator/FileManipulator.csproj [hunk 4]
```diff
@@ -3,16 +3,16 @@
...
+        <LangVersion>9.0</LangVersion>
         <Nullable>enable</Nullable>
         <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
         <GenerateDocumentationFile>true</GenerateDocumentationFile>
     </PropertyGroup>
     <ItemGroup>
-        <ProjectReference Include="..\FileManipulatorAbstractions\FileManipulatorAbstractions.csproj" />
+        <AssemblyAttribute Include="System.Runtime.CompilerServices.InternalsVisibleTo">
+            <_Parameter1>$(AssemblyName).Test</_Parameter1>
+        </AssemblyAttribute>
```
Comment (tautropfli) [inline FileManipulator/FileManipulator.csproj]: This is supported by the .NET 5 SDK:
```suggestion
        <InternalsVisibleTo Include="$(AssemblyName).Test" />
```

File: MetaGenerator/UserInput/UserInputProvider.cs [hunk 5]
```diff
@@ -84,9 +84,15 @@ public string Value(string variableName)
...
+                .FromNullable(stream)
+                .Match(
+                    none: () => throw new Exception("no stream"),
+                    some: s => Option.FromNullable((List<Variable>?)_jsonSerializer.ReadObject(s)));
 
-            return (List<Variable>)_jsonSerializer
-                .ReadObject(stream);
+            return list.Match(
+                none: () => throw new Exception("read object failed"),
+                some: Identity);
```
Comment (tautropfli) [inline MetaGenerator/UserInput/UserInputProvider.cs]:
```suggestion
            return list.GetOrElse(() => throw new Exception("read object failed"));
```

File: Packages.props [hunk 6]
```diff
@@ -1,29 +1,29 @@
...
+		<PackageReference Update="Soltys.ChangeCase" Version="[1.6.0, 2)" />
+		<PackageReference Update="Stubble.Core" Version="[1.9.3, 2)" />
+		<PackageReference Update="System.Collections.Immutable" Version="[5.0.0, 6)" />
+		<PackageReference Update="System.CommandLine" Version="[2.0.0-beta1.20574.7 , 3)" />
+		<PackageReference Update="xunit" Version="[2.4.1, 3)" />
+		<PackageReference Update="xunit.runner.visualstudio" Version="[2.4.3, 3)" />
 	</ItemGroup>
 	<ItemGroup>
-		<PackageReference Update="Messerli.CodeStyle" Version="[1.2.0, 2)" />
+		<PackageReference Update="Messerli.CodeStyle" Version="[1.2.1, 2)" />
```
Comment (tautropfli) [inline Packages.props]: The [2.0.0 rc](https://www.nuget.org/packages/Messerli.CodeStyle/2.0.0-rc.3) provides better compatibility with C# 9 features. Usage: https://github.com/messerli-informatik-ag/code-style

File: Plugins/MesserliOneRepositoryPlugin/TargetFrameworkProvider.cs [hunk 7]
```diff
@@ -108,8 +109,15 @@ private List<DotNetSdk> GetSdks()
...
-                .ReadObject(stream);
+            var list = Option
+                .FromNullable(stream)
+                .Match(
+                    none: () => throw new Exception("no stream"),
+                    some: s => Option.FromNullable((List<DotNetSdk>?)_jsonSerializer.ReadObject(s)));
+
+            return list.Match(
+                none: () => throw new Exception("read object failed"),
+                some: Identity);
```
Comment (tautropfli) [inline Plugins/MesserliOneRepositoryPlugin/TargetFrameworkProvider.cs]:
```suggestion
            return list.GetOrElse(() => throw new Exception("read object failed"));
```

File: VsSolution/Model/Project.cs [hunk 8]
```diff
@@ -1,16 +1,17 @@
...
     public class Project
     {
-        public Project(string projectName, string projectPath, ProjectType.Identifier projectType, Guid? projectGuid)
+        public Project(string projectName, string projectPath, ProjectType.Identifier projectType, Option<Guid> projectGuid = default)
         {
             ProjectName = projectName;
             ProjectPath = projectPath;
             ProjectType = new ProjectType(projectType);
-            ProjectGuid = projectGuid ?? Guid.NewGuid();
+            ProjectGuid = projectGuid.GetOrElse(Guid.NewGuid());
```
Comment (tautropfli) [inline VsSolution/Model/Project.cs]: Does this work?
```suggestion
            ProjectGuid = projectGuid.GetOrElse(Guid.NewGuid);
```

File: global.json [hunk 9]
```diff
@@ -1,8 +1,8 @@
 {
   "sdk": {
-    "version": "3.1.100"
+    "version": "5.0.100"
```
Comment (tautropfli) [inline global.json]: Configure a [`rollForward`](https://docs.microsoft.com/en-us/dotnet/core/tools/global-json?tabs=netcore3x#rollforward) policy. I'd use `latestFeature` or `latestMinor`:

> `latestFeature`: Uses the highest installed feature band and patch level that matches  the requested major and minor with a feature band and patch level that  is greater or equal than the specified value.  If not found, fails.
> `latestMinor`: Uses the highest installed feature band and patch level that matches the requested major and minor with a feature band and patch level that is greater or equal than the specified value.
If not found, fails.

**Review by tautropfli — CHANGES_REQUESTED** [review]

**Comment by tautropfli** [comment]
I think some of the test errors should be fixed if you update the `Microsoft.Build` and the `Microsoft.Build.Locator` packages

File: FileManipulator.Test/GlobalJsonManipulatorTest.cs (hunk 2 above)
Comment (FreeApophis) [reply FileManipulator.Test/GlobalJsonManipulatorTest.cs]: That was resharper :)

Comment (FreeApophis) [reply FileManipulator.Test/GlobalJsonManipulatorTest.cs]: Adressed in 850dd740556684d43fef0fe1b4d1a2b2b3402f2f

File: .github/workflows/dotnetcore.yml (hunk 1 above)
Comment (FreeApophis) [reply .github/workflows/dotnetcore.yml]: Adressed in 6c9d69126b7cd3825d4b1afe219814e5f95b1f51

File: Packages.props (hunk 6 above)
Comment (FreeApophis) [reply Packages.props]: Addressed in 0970d02e67cdc9ea2a2de6ed4f19eab51fcb0482

File: VsSolution/Model/Project.cs (hunk 8 above)
Comment (FreeApophis) [reply VsSolution/Model/Project.cs]: It indeed does

File: global.json (hunk 9 above)
Comment (FreeApophis) [reply global.json]: Set to latestFeature in 616b415c79cf54eab14a0fa3ae4bb5d3a2362cae

File: FileManipulatorAbstractions/GlobalJsonModificationBuilder.cs [hunk 10]
```diff
@@ -21,14 +22,14 @@ private GlobalJsonModificationBuilder(IImmutableList<MsBuildSdk> msBuildSdksToAd
 
         [Pure]
         public GlobalJsonModificationBuilder AddMsBuildSdk(MsBuildSdk sdk)
-            => ShallowClone(sdksToAdd: _msBuildSdksToAdd.Add(sdk));
+            => ShallowClone(sdksToAdd: Option.Some(_msBuildSdksToAdd.Add(sdk)));
```
Comment (tautropfli) [inline FileManipulatorAbstractions/GlobalJsonModificationBuilder.cs:25]: Use the implicit conversion here

File: MetaGeneratorAbstractions/Json/Variable.cs [hunk 11]
```diff
@@ -21,8 +21,11 @@ public class Variable
         public List<SelectionValue>? SelectionValues { get; set; }
 
         [DataMember]
-        public List<string> Validations { get; set; } = new List<string>();
+        public List<string> Validations { get; set; } = new();
 
-        public VariableType GetVariableType() => Enum.Parse<VariableType>(Type);
+        public VariableType GetVariableType()
+            => Type is null
+            ? throw new Exception("Type is null")
```
Comment (tautropfli) [inline MetaGeneratorAbstractions/Json/Variable.cs]: Indentation

File: global.json [hunk 12]
```diff
@@ -1,8 +1,9 @@
 {
-  "sdk": {
-    "version": "3.1.100"
+    "sdk": {
+        "version": "5.0.100",
+        "rollForward": "latestFeature"
   },
   "msbuild-sdks": {
     "Microsoft.Build.CentralPackageVersions": "2.0.52"
```
Comment (tautropfli) [inline global.json]:
```suggestion
    "Microsoft.Build.CentralPackageVersions": "2.0.79"
```

**Review by tautropfli — CHANGES_REQUESTED** [review]

File: FileManipulatorAbstractions/GlobalJsonModificationBuilder.cs (hunk 10 above)
Comment (FreeApophis) [reply FileManipulatorAbstractions/GlobalJsonModificationBuilder.cs:25]: That does not work we work here on the `IImmutableList<T>`

File: MetaGeneratorAbstractions/Json/Variable.cs (hunk 11 above)
Comment (FreeApophis) [reply MetaGeneratorAbstractions/Json/Variable.cs]: Addressed in 5e5b56f2

**Review by tautropfli — APPROVED** [review]

**Pull request merged**